client/asset/eth/eth.go

```diff
@@ -593,101 +824,284 @@ func (c *coin) Value() uint64 {
+	//
+	// ethToLock += biggestUint64(g.oneRefund, g.oneRedeem) * cfg.MaxFeeRate * ord.MaxSwapCount
+
+	ethWallet := w.ethWallet()
+	var success bool
+	var coin asset.Coin
+	if w.assetID == BipID {
+		ethToLock += ord.Value
+		coin = w.createFundingCoin(ethToLock)
+	} else {
+		if err := w.lockFunds(ord.Value, initiationReserve); err != nil {
+			return nil, nil, fmt.Errorf("error locking token funds: %v", err)
+		}
+		defer func() {
+			if !success {
+				w.unlockFunds(ord.Value, initiationReserve)
+			}
+		}()
+
+		coin = w.createTokenFundingCoin(ord.Value, ethToLock)
+	}
+
+	if err := ethWallet.lockFunds(ethToLock, initiationReserve); err != nil {
 		return nil, nil, err
 	}
 
-	return coins, []dex.Bytes{nil}, nil
+	success = true
+	return asset.Coins{coin}, []dex.Bytes{nil}, nil
+}
+
+// gasEstimates are estimates of gas required for operations involving a swap
+// combination of (swap asset, redeem asset, # lots).
+type gasEstimate struct {
+	// The embedded are best calculated values for Swap gases, and redeem costs
+	// IF the redeem asset is a fee-family asset. Otherwise Redeem and RedeemAdd
+	// are zero.
+	dexeth.Gases
+	// Additional fields are based on live estimates of the swap.
+	oneGas, nGas, nSwap, nRedeem, nRefund uint64
+}
+
+// initGasEstimate gets the best available gas estimate for n initiations. A
+// live estimate is checked against the server's configured values and our own
+// known values and errors or logs generated in certain cases.
+func (w *AssetWallet) initGasEstimate(n int, dexSwapCfg, dexRedeemCfg *dex.Asset) (est *gasEstimate, err error) {
+	est = new(gasEstimate)
+
+	g := w.gases(dexSwapCfg.Version)
+	if g == nil {
+		return nil, fmt.Errorf("no gas table")
+	}
+
+	est.Refund = g.Refund
+
+	est.Swap, est.nSwap, err = w.swapGas(n, dexSwapCfg)
+	if err != nil {
+		return nil, fmt.Errorf("error calculating swap gas: %w", err)
+	}
+
+	est.oneGas = est.Swap + est.Redeem
+	est.nGas = est.nSwap + est.nRedeem
+
+	if redeemW := w.wallet(dexRedeemCfg.ID); redeemW != nil {
+		est.Redeem, est.nRedeem, err = redeemW.redeemGas(n, dexRedeemCfg)
+		if err != nil {
+			return nil, fmt.Errorf("error calculating fee-family redeem gas: %w", err)
+		}
+	}
+
+	return
+}
+
+// swapGas estimates gas for a number of initiations. swapGas will error if we
+// cannot get a live estimate from the contractor, which will happen if the
+// wallet has no balance.
+func (w *AssetWallet) swapGas(n int, dexCfg *dex.Asset) (oneSwap, nSwap uint64, err error) {
+	oneSwap = dexCfg.SwapSize
+
+	g := w.gases(dexCfg.Version)
+	if g == nil {
+		return 0, 0, fmt.Errorf("no gases known for %d version %d", w.assetID, dexCfg.Version)
+	}
+
+	// Warn of mismatch in gas estimates.
+	if g.Swap > oneSwap {
+		w.log.Warnf("Server has a lower configured swap gas value, which is odd")
+		oneSwap = g.Swap
+	}
+
+	// Not gonna worry about accuracy for nSwap. It's never used where it
+	// matters, just for estimates.
+	nSwap = oneSwap + uint64(n-1)*g.SwapAdd
```

Comment:
What do you mean by worry about accuracy?  As in just using the hard coded and padded estimates for this contract version?